### ISSUE TITLE
alpha.08 : Error inserting batch: no such table: location

### ISSUE THREAD

**Issue by anymos** [mauron85/cordova-plugin-background-geolocation#163]
#

## Your Environment

* Plugin version:alpha-8
* Platform: iOS
* OS version:9.3
* Device manufacturer and model:ipad pro
* Cordova version (`cordova -v`):Apache Cordova native platform version 4.1.0 is starting.
* Cordova platform version (`cordova platform ls`) : Apache Cordova native platform version 4.1.0 is starting.
* Plugin configuration options: 
* Link to your project:

## Context
running on the device 

2016-09-14 11:24:11.333 [341:47001] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///private/var/mobile/Containers/Data/Application/B636752A-5407-414B-B07C-1F719A24E392/Library/Cookies/Cookies.binarycookies
2016-09-14 11:24:11.408 [341:47001] Apache Cordova native platform version 4.1.0 is starting.
2016-09-14 11:24:11.408 [341:47001] Multi-tasking -> Device: YES, App: YES
2016-09-14 11:24:11.460 [341:47001] Using WKWebView
2016-09-14 11:24:11.460 [341:47001] [CDVTimer][handleopenurl] 0.048995ms
2016-09-14 11:24:11.462 [341:47001] [CDVTimer][intentandnavigationfilter] 1.370013ms
2016-09-14 11:24:11.462 [341:47001] [CDVTimer][gesturehandler] 0.051975ms
2016-09-14 11:24:11.514 [341:47001] Serving asset bundle version: a8feb266d733bd600511eb7d198119c35395361a
[INFO] GCDWebServer started on port 12496 and reachable at http://localhost:12496/
2016-09-14 11:24:11.515 [341:47001] [CDVTimer][webapplocalserver] 53.530037ms
2016-09-14 11:24:11.525 [341:47001] [CDVTimer][statusbar] 9.514987ms
2016-09-14 11:24:11.549 [341:47001] [CDVTimer][splashscreen] 23.597002ms
2016-09-14 11:24:11.570 [341:47001] [CDVTimer][backgroundgeolocation] 20.833015ms
2016-09-14 11:24:11:569 [341:47001] LocationManager iOS9 detected
2016-09-14 11:24:11.573 [341:47001] [CDVTimer][keyboard] 3.112972ms
2016-09-14 11:24:11.574 [341:47001] [CDVTimer][localstorage] 0.536025ms
2016-09-14 11:24:11.574 [341:47001] [CDVTimer][TotalPluginStartup] 113.664031ms
2016-09-14 11:24:12:016 [341:47001] LocationManager didChangeAuthorizationStatus 0
2016-09-14 11:24:43:916 [341:47308] LocationManager stop
2016-09-14 11:24:43.920 [341:47283] DB Error: 1 "no such table: __META"
2016-09-14 11:24:43.920 [341:47283] DB Query: SELECT db_version FROM __META
2016-09-14 11:24:43.920 [341:47283] DB Path: /var/mobile/Containers/Data/Application/B636752A-5407-414B-B07C-1F719A24E392/Library/cordova_bg_geolocation.db
2016-09-14 11:24:43.935 [341:47283] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
2016-09-14 11:24:43.936 [341:47283] Current db version: 0
2016-09-14 11:24:43.936 [341:47283] Error inserting batch: no such table: location
2016-09-14 11:24:43.936 [341:47283] DROP TABLE location failed code: 1: message: no such table: location
2016-09-14 11:24:44.425 [341:47001] THREAD WARNING: ['GoogleMaps'] took '350.446777' ms. Plugin should use a background thread.

## Expected Behavior
No db error

## Actual Behavior
db error

## Steps to Reproduce

1.
2.
3.
4.

## Context
testing the backgroudn mode in ios

## Debug logs
2016-09-14 11:24:11.333 [341:47001] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///private/var/mobile/Containers/Data/Application/B636752A-5407-414B-B07C-1F719A24E392/Library/Cookies/Cookies.binarycookies
2016-09-14 11:24:11.408 [341:47001] Apache Cordova native platform version 4.1.0 is starting.
2016-09-14 11:24:11.408 [341:47001] Multi-tasking -> Device: YES, App: YES
2016-09-14 11:24:11.460 [341:47001] Using WKWebView
2016-09-14 11:24:11.460 [341:47001] [CDVTimer][handleopenurl] 0.048995ms
2016-09-14 11:24:11.462 [341:47001] [CDVTimer][intentandnavigationfilter] 1.370013ms
2016-09-14 11:24:11.462 [341:47001] [CDVTimer][gesturehandler] 0.051975ms
2016-09-14 11:24:11.514 [341:47001] Serving asset bundle version: a8feb266d733bd600511eb7d198119c35395361a
[INFO] GCDWebServer started on port 12496 and reachable at http://localhost:12496/
2016-09-14 11:24:11.515 [341:47001] [CDVTimer][webapplocalserver] 53.530037ms
2016-09-14 11:24:11.525 [341:47001] [CDVTimer][statusbar] 9.514987ms
2016-09-14 11:24:11.549 [341:47001] [CDVTimer][splashscreen] 23.597002ms
2016-09-14 11:24:11.570 [341:47001] [CDVTimer][backgroundgeolocation] 20.833015ms
2016-09-14 11:24:11:569 [341:47001] LocationManager iOS9 detected
2016-09-14 11:24:11.573 [341:47001] [CDVTimer][keyboard] 3.112972ms
2016-09-14 11:24:11.574 [341:47001] [CDVTimer][localstorage] 0.536025ms
2016-09-14 11:24:11.574 [341:47001] [CDVTimer][TotalPluginStartup] 113.664031ms
2016-09-14 11:24:12:016 [341:47001] LocationManager didChangeAuthorizationStatus 0
2016-09-14 11:24:43:916 [341:47308] LocationManager stop
2016-09-14 11:24:43.920 [341:47283] DB Error: 1 "no such table: __META"
2016-09-14 11:24:43.920 [341:47283] DB Query: SELECT db_version FROM __META
2016-09-14 11:24:43.920 [341:47283] DB Path: /var/mobile/Containers/Data/Application/B636752A-5407-414B-B07C-1F719A24E392/Library/cordova_bg_geolocation.db
2016-09-14 11:24:43.935 [341:47283] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
2016-09-14 11:24:43.936 [341:47283] Current db version: 0
2016-09-14 11:24:43.936 [341:47283] Error inserting batch: no such table: location
2016-09-14 11:24:43.936 [341:47283] DROP TABLE location failed code: 1: message: no such table: location
2016-09-14 11:24:44.425 [341:47001] THREAD WARNING: ['GoogleMaps'] took '350.446777' ms. Plugin should use a background thread.

**Comment by mauron85** [comment]
Did you upgrade from previous version of plugin? If yes, which version?

**Comment by mauron85** [comment]
```
2016-09-14 11:24:43.936 Mydelivery Driver[341:47283] Current db version: 0
```

SQLite seems not created properly. Can you try to uninstall app completely and install fresh?

**Comment by mauron85** [comment]
That's strange. Your current db version is 0, but __META table doesn't exists. And db version is stored in __META table.

**Comment by anymos** [comment]
I just saw that you release a new version, I am going to test with it, as it s late already in Thailand, I will keep you posted tomorow monring

**Comment by mauron85** [comment]
there will be no change regarding this

**Comment by anymos** [comment]
Right, no change unfortunatly

same log :

2016-09-14 20:34:10.245 Mydelivery Driver[437:126410] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///private/var/mobile/Containers/Data/Application/9EEA020C-5A8B-40C2-9B85-A0A51D0C6771/Library/Cookies/Cookies.binarycookies
2016-09-14 20:34:10.343 Mydelivery Driver[437:126410] Apache Cordova native platform version 4.1.0 is starting.
2016-09-14 20:34:10.343 Mydelivery Driver[437:126410] Multi-tasking -> Device: YES, App: YES
2016-09-14 20:34:10.390 Mydelivery Driver[437:126410] Using WKWebView
2016-09-14 20:34:10.390 Mydelivery Driver[437:126410] [CDVTimer][handleopenurl] 0.050008ms
2016-09-14 20:34:10.393 Mydelivery Driver[437:126410] [CDVTimer][intentandnavigationfilter] 2.440035ms
2016-09-14 20:34:10.393 Mydelivery Driver[437:126410] [CDVTimer][gesturehandler] 0.063002ms
2016-09-14 20:34:10.440 Mydelivery Driver[437:126410] Serving asset bundle version: 6e013d2cd5662d8b2b7bc2a47c701bc3da334ed2
[INFO] GCDWebServer started on port 12496 and reachable at http://localhost:12496/
2016-09-14 20:34:10.441 Mydelivery Driver[437:126410] [CDVTimer][webapplocalserver] 48.396945ms
2016-09-14 20:34:10.449 Mydelivery Driver[437:126410] [CDVTimer][statusbar] 7.749975ms
2016-09-14 20:34:10.465 Mydelivery Driver[437:126410] [CDVTimer][splashscreen] 15.592992ms
2016-09-14 20:34:10.487 Mydelivery Driver[437:126410] [CDVTimer][backgroundgeolocation] 21.620989ms
2016-09-14 20:34:10:486 Mydelivery Driver[437:126410] LocationManager iOS9 detected
2016-09-14 20:34:10.490 Mydelivery Driver[437:126410] [CDVTimer][keyboard] 3.141999ms
2016-09-14 20:34:10.491 Mydelivery Driver[437:126410] [CDVTimer][localstorage] 0.554979ms
2016-09-14 20:34:10.491 Mydelivery Driver[437:126410] [CDVTimer][TotalPluginStartup] 100.650012ms
2016-09-14 20:34:10:920 Mydelivery Driver[437:126410] LocationManager didChangeAuthorizationStatus 0
2016-09-14 20:35:15.340 Mydelivery Driver[437:126410] Start downloading asset manifest from: manifest.json -- http://192.168.8.134:3000/__cordova/
2016-09-14 20:35:15.450 Mydelivery Driver[437:126867] Downloaded asset manifest for version: 156f2d79bf1bf3e5c2cf2b3d083d58aacf534852
2016-09-14 20:35:15.479 Mydelivery Driver[437:126882] Start downloading assets from bundle with version: 156f2d79bf1bf3e5c2cf2b3d083d58aacf534852
2016-09-14 20:35:15.629 Mydelivery Driver[437:126884] [CDVTimer][assetBundleDownload] 149.667978ms
2016-09-14 20:35:15.631 Mydelivery Driver[437:126882] Finished downloading new asset bundle version: 156f2d79bf1bf3e5c2cf2b3d083d58aacf534852
2016-09-14 20:35:15.656 Mydelivery Driver[437:126410] Serving asset bundle version: 156f2d79bf1bf3e5c2cf2b3d083d58aacf534852
2016-09-14 20:35:28:819 Mydelivery Driver[437:126886] LocationManager stop
2016-09-14 20:35:28.822 Mydelivery Driver[437:126879] DB Error: 1 "no such table: __META"
2016-09-14 20:35:28.822 Mydelivery Driver[437:126879] DB Query: SELECT db_version FROM __META
2016-09-14 20:35:28.823 Mydelivery Driver[437:126879] DB Path: /var/mobile/Containers/Data/Application/9EEA020C-5A8B-40C2-9B85-A0A51D0C6771/Library/cordova_bg_geolocation.db
2016-09-14 20:35:28.837 Mydelivery Driver[437:126879] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
2016-09-14 20:35:28.837 Mydelivery Driver[437:126879] Current db version: 0
2016-09-14 20:35:28.837 Mydelivery Driver[437:126879] Error inserting batch: no such table: location
2016-09-14 20:35:28.837 Mydelivery Driver[437:126879] DROP TABLE location failed code: 1: message: no such table: location
2016-09-14 20:35:29.321 Mydelivery Driver[437:126410] THREAD WARNING: ['GoogleMaps'] took '284.207031' ms. Plugin should use a background thread.
2016-09-14 20:35:48:738 Mydelivery Driver[437:126410] LocationManager didChangeAuthorizationStatus 4
2016-09-14 20:35:57:152 Mydelivery Driver[437:127084] LocationManager will start: 0
2016-09-14 20:35:57:153 Mydelivery Driver[437:127084] LocationManager switchMode 1
2016-09-14 20:35:57:164 Mydelivery Driver[437:126410] LocationManager didUpdateLocations (operationMode: 1)
2016-09-14 20:35:57:164 Mydelivery Driver[437:126410] Location age 0.009972
2016-09-14 20:35:57:167 Mydelivery Driver[437:126410] LocationManager didUpdateLocations (operationMode: 1)
2016-09-14 20:35:57:167 Mydelivery Driver[437:126410] Location age 0.012500
2016-09-14 20:35:57:167 Mydelivery Driver[437:126410] Better location found: Location: id=0 time=5098027360 lat=13.69327831785719 lon=100.6210846954536 accu=65 aaccu=10 speed=-1 bear=-1 alt=0.1314567029476166 type=current

**Comment by anymos** [comment]
in case the other plugins I am using 

cordova-plugin-nativestorage@2.0.2
cordova-plugin-splashscreen@3.2.1
cordova-plugin-mauron85-background-geolocation@2.2.0
com.googlemaps.ios@file://../../cordovaDev/cordova-plugin-googlemaps-sdk
cordova-plugin-googlemaps@file://../../cordovaDev/cordova-plugin-googlemaps
cordova-plugin-statusbar@https://github.com/aegliv/cordova-plugin-statusbar.git#8467e8e51c9747f3fcdf0c611093486bc984b3ce
ionic-plugin-keyboard@2.2.1
cordova-plugin-device@1.1.2
cordova-plugin-market@1.1.3
cordova-plugin-appavailability@0.4.2
cordova-plugin-contacts@2.1.0

I am goign to remove the native storage just in case

**Comment by anymos** [comment]
could it have happen, that I remove the app while the service is still running, producing a corrupted db ?

**Comment by anymos** [comment]
Same error, I will try on an iphone 6 tomorow. Anything I could do to help finding the root cause of this one ?

**Comment by mauron85** [comment]
Did you uninstall app before upgrading?

**Comment by anymos** [comment]
Yes I uninstall the app. Could it be a problem if 2 app are using the same plugin ? it should not right ?

**Comment by anymos** [comment]
Good morning, to make sure that it was not coming from my app it self, I have reproduced the issue with the BG sample app, logs are showing the same error 

2016-09-15 07:16:56.883 BG Example[382:51346] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///private/var/mobile/Containers/Data/Application/0C46E079-91AF-4E4A-90CB-99E77772AF5F/Library/Cookies/Cookies.binarycookies
2016-09-15 07:16:56.962 BG Example[382:51346] Apache Cordova native platform version 4.1.1 is starting.
2016-09-15 07:16:56.962 BG Example[382:51346] Multi-tasking -> Device: YES, App: YES
2016-09-15 07:16:56.967 BG Example[382:51346] 

Started backup to iCloud! Please be careful.
Your application might be rejected by Apple if you store too much data.
For more information please read "iOS Data Storage Guidelines" at:
https://developer.apple.com/icloud/documentation/data-storage/
To disable web storage backup to iCloud, set the BackupWebStorage preference to "local" in the Cordova config.xml file

2016-09-15 07:16:57.073 BG Example[382:51346] Using UIWebView
2016-09-15 07:16:57.075 BG Example[382:51346] [CDVTimer][handleopenurl] 0.063002ms
2016-09-15 07:16:57.076 BG Example[382:51346] [CDVTimer][intentandnavigationfilter] 1.057029ms
2016-09-15 07:16:57.076 BG Example[382:51346] [CDVTimer][gesturehandler] 0.050008ms
2016-09-15 07:16:57.101 BG Example[382:51346] [CDVTimer][backgroundgeolocation] 25.273979ms
2016-09-15 07:16:57.101 BG Example[382:51346] [CDVTimer][TotalPluginStartup] 26.848018ms
2016-09-15 07:16:57:101 BG Example[382:51346] LocationManager iOS9 detected
2016-09-15 07:16:57:765 BG Example[382:51346] LocationManager didChangeAuthorizationStatus 0
2016-09-15 07:16:57.806 BG Example[382:51346] Resetting plugins due to page load.
2016-09-15 07:16:58.210 BG Example[382:51490] DB Error: 1 "no such table: __META"
2016-09-15 07:16:58.210 BG Example[382:51490] DB Query: SELECT db_version FROM __META
2016-09-15 07:16:58.210 BG Example[382:51490] DB Path: /var/mobile/Containers/Data/Application/0C46E079-91AF-4E4A-90CB-99E77772AF5F/Library/cordova_bg_geolocation.db
2016-09-15 07:16:58.225 BG Example[382:51490] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
2016-09-15 07:16:58.225 BG Example[382:51490] Current db version: 0
2016-09-15 07:16:58.225 BG Example[382:51490] Error inserting batch: no such table: location
2016-09-15 07:16:58.226 BG Example[382:51490] DROP TABLE location failed code: 1: message: no such table: location
2016-09-15 07:16:58:228 BG Example[382:51465] Config: distanceFilter=50 stationaryRadius=50 desiredAccuracy=10 activityType=AutomotiveNavigation isDebugging=1 stopOnTerminate=0 url=http://192.168.81.15:3000/locations httpHeaders={
    "X-FOO" = bar;
} pauseLocationUpdates=0
2016-09-15 07:16:58.229 BG Example[382:51465] +backgroundSessionConfiguration: is deprecated. Please use +backgroundSessionConfigurationWithIdentifier: instead
2016-09-15 07:16:59.749 BG Example[382:51346] Finished load of: file:///var/mobile/Containers/Bundle/Application/DF9FC143-BB42-44E4-8C84-C885CFFC2CFA/BG%20Example.app/www/index.ios.html

**Comment by anymos** [comment]
Good morning/evening

I am happy I find a easy way to reproduce the issue in the ios simulator, I believe that it would be helpful. Could you kindly tell me if you are able to reproduce it on your side ?

Step to reproduce :

1. clone the latest sample 
2. cordova platform add ios
3. cd platforms/ios
4. double click on BG Example.xcodeproj
5. select a device never used
6. run the build, ( let the popup asking for the notification on the screen )


logs are bellow :


2016-09-15 07:23:58.152 BG Example[2406:26141] DiskCookieStorage changing policy from 2 to 0, cookie file: file:///Users/tantr/Library/Developer/CoreSimulator/Devices/7414366B-EEA8-4B5C-958F-F792FE6118E9/data/Containers/Data/Application/E1273C2B-7E19-4240-A1EC-4A95B5F3DF2C/Library/Cookies/com.marianhello.bgexampleapp.binarycookies
2016-09-15 07:23:58.171 BG Example[2406:26141] Apache Cordova native platform version 4.1.1 is starting.
2016-09-15 07:23:58.171 BG Example[2406:26141] Multi-tasking -> Device: YES, App: YES
2016-09-15 07:23:58.174 BG Example[2406:26141] 

Started backup to iCloud! Please be careful.
Your application might be rejected by Apple if you store too much data.
For more information please read "iOS Data Storage Guidelines" at:
https://developer.apple.com/icloud/documentation/data-storage/
To disable web storage backup to iCloud, set the BackupWebStorage preference to "local" in the Cordova config.xml file

2016-09-15 07:23:58.197 BG Example[2406:26141] Using UIWebView
2016-09-15 07:23:58.198 BG Example[2406:26141] [CDVTimer][handleopenurl] 0.039995ms
2016-09-15 07:23:58.199 BG Example[2406:26141] [CDVTimer][intentandnavigationfilter] 0.894010ms
2016-09-15 07:23:58.199 BG Example[2406:26141] [CDVTimer][gesturehandler] 0.032961ms
2016-09-15 07:23:58.204 BG Example[2406:26141] [CDVTimer][backgroundgeolocation] 5.165994ms
2016-09-15 07:23:58.204 BG Example[2406:26141] [CDVTimer][TotalPluginStartup] 6.593049ms
2016-09-15 07:23:58:203 BG Example[2406:26141] LocationManager iOS9 detected
2016-09-15 07:23:58:521 BG Example[2406:26141] LocationManager didChangeAuthorizationStatus 0
2016-09-15 07:23:58.542 BG Example[2406:26141] Resetting plugins due to page load.
2016-09-15 07:23:58.796 BG Example[2406:26217] DB Error: 1 "no such table: __META"
2016-09-15 07:23:58.796 BG Example[2406:26217] DB Query: SELECT db_version FROM __META
2016-09-15 07:23:58.796 BG Example[2406:26217] DB Path: /Users/tantr/Library/Developer/CoreSimulator/Devices/7414366B-EEA8-4B5C-958F-F792FE6118E9/data/Containers/Data/Application/E1273C2B-7E19-4240-A1EC-4A95B5F3DF2C/Library/cordova_bg_geolocation.db
2016-09-15 07:23:58.797 BG Example[2406:26217] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
2016-09-15 07:23:58.798 BG Example[2406:26217] Current db version: 0
2016-09-15 07:23:58.798 BG Example[2406:26217] Error inserting batch: no such table: location
2016-09-15 07:23:58.798 BG Example[2406:26217] DROP TABLE location failed code: 1: message: no such table: location
2016-09-15 07:23:58:811 BG Example[2406:26225] Config: distanceFilter=50 stationaryRadius=50 desiredAccuracy=10 activityType=AutomotiveNavigation isDebugging=1 stopOnTerminate=0 url=http://192.168.81.15:3000/locations httpHeaders={
    "X-FOO" = bar;
} pauseLocationUpdates=0
2016-09-15 07:23:58.812 BG Example[2406:26225] +backgroundSessionConfiguration: is deprecated. Please use +backgroundSessionConfigurationWithIdentifier: instead
2016-09-15 07:24:00.313 BG Example[2406:26141] Finished load of: file:///Users/tantr/Library/Developer/CoreSimulator/Devices/7414366B-EEA8-4B5C-958F-F792FE6118E9/data/Containers/Bundle/Application/A19B87A9-1E44-405E-ADD2-0799585E9C7C/BG%20Example.app/www/index.ios.html

**Comment by mauron85** [comment]
Ok, you're right. But actually there is no problem with it. I just forgot, what is it doing. Just for future reference.

```
2016-09-15 07:23:58.796 BG Example[2406:26217] DB Error: 1 "no such table: __META"
```
`openOrCreateDatabase` is invoking `getVersion` method in SQLiteOpenHelper is called to get version of db, but db not exists yet. This is expected and will return -1, which means to create db for the first time (this message is only show on first run).

```
2016-09-15 07:23:58.797 BG Example[2406:26217] INSERT INTO __META (id,db_version,created,last_updated) VALUES (1,0,datetime(),null) failed code: 0: message: not an error
```

this is due small bug in error handling, will fix this, but does'n effect functionally in any way.

```
2016-09-15 07:23:58.798 BG Example[2406:26217] Current db version: 0
2016-09-15 07:23:58.798 BG Example[2406:26217] Error inserting batch: no such table: location
2016-09-15 07:23:58.798 BG Example[2406:26217] DROP TABLE location failed code: 1: message: no such table: location
```
this is to handle situation when legacy version of plugin with no db version was installed before. there is already comment in GeolocationOpenHelper explaining why : '// because of some legacy code we have to drop table'. This can be also ignored.

So there is no problem and those messages can be ignored.

**Comment by mauron85** [comment]
will try to reduce number of those message in future version though

**Comment by mauron85** [comment]
fixed in 2.2.1

**Comment by anymos** [comment]
Understood, No words to tell you how much I appreciate it your feedback on this one !

Thank you so much !

I am working on the second issue now, I will be able to post my findings today hopefully.

Thanks again
